### ISSUE TITLE
malloc fail when cpu_pages is not enough, a global memory area is a better solution

### ISSUE THREAD

**Issue by xuguruogu** [scylladb/seastar#347]
I think a global memory section is needed, just like jemalloc.

**Comment by avikivity** [comment]
You mean jemalloc can't run out of memory/

**Comment by xuguruogu** [comment]
Seastar simply throw an error when the shard of the CPU is out of memory. If it try to allocate memory in global section after fail in local shard, the operation with temporary large memory requirement will not fail.

**Comment by avikivity** [comment]
Seastar is designed around partitioning memory across shards, moving to a global allocator would break that. Seastar applications should similarly break up their work (and any temporary allocations) so each shard allocates only its own memory.

**Comment by nyh** [comment]
On Tue, Oct 17, 2017 at 4:41 PM, Avi Kivity <notifications@github.com>
wrote:

> Seastar is designed around partitioning memory across shards, moving to a
> global allocator would break that. Seastar applications should similarly
> break up their work (and any temporary allocations) so each shard allocates
> only its own memory.
>
>
In theory, one may want the memory to be partitioned between shards, but
not at equal amounts - with a shard that temporarily needs more memory to
use more memory than other shards.

However, in practice, I think if we were to allow such memory movement
between shards (and we don't, currently), it would not be very useful:

First of all, most complex servers do a lot of things in parallel, and
assuming these "things" are distributed to shards quite randomly and
evenly, you should not have different memory needs for the different
shards.  If the work is not evenly distributed between shards, you may be
doing something wrong in designing your application.

Second, many if not most Seastar applications would probably want to take
most of the machine's memory and use it for things like caching. This means
when a shard runs out of memory, what it should do isn't to take more
memory from another shard, but rather evict stuff it keeps cached. In the
Scylla application we have a mechanism - LSA (log structured allocation) -
for helping building such caches and automatically evict from them
(supporting also compaction of holes), but unfortunately it is not well
documented and not (yet) part of Seastar's core.

Finally, note that modern multi-socket machines are NUMA, i.e., some memory
is closer to some of the cores. On such machines, it is inefficient to
re-assign a chunk of memory to a core on a different socket (although it is
ok to reassign it to one of the other cores in the same socket), making the
ability to reassign memory even less appealing.

**Comment by xuguruogu** [comment]
> First of all, most complex servers do a lot of things in parallel, and
assuming these "things" are distributed to shards quite randomly and
evenly, you should not have different memory needs for the different
shards.  If the work is not evenly distributed between shards, you may be
doing something wrong in designing your application.

We use shared memory for cache. When start the seastar application, less memory is allocated for each shard. In a case we merge two large set, the shard memory allocation operation failed. I think local memory for each shards with a global memory section may be helpful in this case.

**Comment by xuguruogu** [comment]
Maybe give us a choice to select memory allocator is also a good solution.

**Comment by xuguruogu** [comment]
I realize what goes wrong. If I use shared memory for cache, the seastar application still split almost whole memory into shards using mmap. As the memory page do not return back to system when accessed, the overall memory declared may exceed total physical memory and thus be killed. 

I noticed that with DEFAULT_ALLOCATOR option, seastar can use libc malloc as memory allocator, but jemalloc or tcmalloc seam to be better one. So could seastar offer jemalloc or tcmalloc as an alternative memory allocator?

**Comment by nyh** [comment]
@xuguruogu indeed, as you noticed, by default Seastar does take most of the machine's memory and split it equally among the shards because, as I explained earlier, we think this is the best way to do things. However, you can override this default with the "-m" option to the Seastar application (which says how much total memory the application will take for Seastar's memory) or "--reserve-memory" (which says how much memory to LEAVE and not take for Seastar). You can then allocate the memory which Scylla didn't take using mmap(), as you wished.

Defining DEFAULT_ALLOCATOR simply means that Seastar does not redefine malloc() or free() at all. You can still redefine those to use a different library, if you wish to do so.
